src/types.ts

```diff
@@ -0,0 +1,37 @@
+import { Matcher } from './compile-matcher';
+import { OBJECT_REST } from './rest';
+
+// The __foo__ indirection is required to keep TS happy with our sneaky recursive types
+export type CapturedValues<T> = FindCapturedValues<T>['__captured__'];
+export type MatchedValue<T> = FindMatchedValue<T>['__matched__'];
+
+type FindCapturedValues<T>
```

Comment:
my brain hurts

Reply:
added some comments to this file, but let's definitely go through it tomorrow